larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,240 @@
+/**
+ *  @file   larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc
+ *
+ *  @brief  Implementation of the lar discrete probability vector class
+ *
+ *  $Log: $
+ */
+
+#include "larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h"
+
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+
+namespace lar_content
+{
+
+template <typename TX, typename TY>
+DiscreteProbabilityVector::DiscreteProbabilityVector(InputData<TX, TY> const &inputData, const TX xUpperBound, 
+    const bool useWidths) :
+    m_xUpperBound(static_cast<float>(xUpperBound)),
+    m_useWidths(useWidths),
+    m_discreteProbabilityData(this->InitialiseDiscreteProbabilityData(inputData))
+{
+    this->VerifyCompleteData();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityVector(DiscreteProbabilityVector const &discreteProbabilityVector,
+    std::mt19937 &randomNumberGenerator) :
+    m_xUpperBound(discreteProbabilityVector.m_xUpperBound),
+    m_useWidths(discreteProbabilityVector.m_useWidths),
+    m_discreteProbabilityData(this->RandomiseDiscreteProbabilityData(discreteProbabilityVector, randomNumberGenerator))
+{
+    this->VerifyCompleteData();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityVector(DiscreteProbabilityVector const &discreteProbabilityVector,
+    ResamplingPoints const &resamplingPoints) :
+    m_xUpperBound(discreteProbabilityVector.m_xUpperBound),
+    m_useWidths(discreteProbabilityVector.m_useWidths),
+    m_discreteProbabilityData(this->ResampleDiscreteProbabilityData(discreteProbabilityVector, resamplingPoints))
+{
+    this->VerifyCompleteData();
+}
+
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float DiscreteProbabilityVector::EvaluateCumulativeProbability(const float x) const
+{
+    if (x - m_discreteProbabilityData.back().GetX() > std::numeric_limits<float>::epsilon())
+        return 1.f;
+
+    if (x - m_discreteProbabilityData.front().GetX() < std::numeric_limits<float>::epsilon())
+        return 0.f;
+
+    for (size_t iDatum = 1; iDatum < m_discreteProbabilityData.size(); ++iDatum)
+    {
+        if (x - m_discreteProbabilityData.at(iDatum).GetX() > std::numeric_limits<float>::epsilon())
+            continue;
+
+        float xLow(m_discreteProbabilityData.at(iDatum-1).GetX());
+        float yLow(m_discreteProbabilityData.at(iDatum-1).GetCumulativeDatum());
+        float xHigh(m_discreteProbabilityData.at(iDatum).GetX());
+        float yHigh(m_discreteProbabilityData.at(iDatum).GetCumulativeDatum());
```

Comment:
const?

Reply:
Got em.  There was also numerous other non-const vars in there that I've got as well.  Thanks very much!